clients/src/main/java/org/apache/kafka/clients/producer/TransactionState.java

```diff
@@ -132,4 +414,332 @@ public synchronized void incrementSequenceNumber(TopicPartition topicPartition,
+                handleResponse(response.responseBody());
+            } else {
+                if (result != null) {
+                    result.setError(Errors.UNKNOWN.exception());
+                    result.done();
+                } else {
+                    throw Errors.UNKNOWN.exception();
+                }
+            }
+        }
+
+        public abstract void handleResponse(AbstractResponse responseBody);
+
+        public abstract void reenqueue();
+    }
+
+    private class InitPidCallback extends TransactionalRequestCallBack {
+
+        InitPidCallback(TransactionalRequestResult result) {
+           super(result);
+        }
+
+        @Override
+        public void handleResponse(AbstractResponse responseBody) {
+            InitPidResponse initPidResponse = (InitPidResponse) responseBody;
+            Errors error = initPidResponse.error();
+            if (error == Errors.NONE) {
+                setPidAndEpoch(initPidResponse.producerId(), initPidResponse.epoch());
+                isInitializing = false;
+            } else if (error == Errors.NOT_COORDINATOR || error == Errors.COORDINATOR_NOT_AVAILABLE) {
+                needsCoordinator(FindCoordinatorRequest.CoordinatorType.TRANSACTION, transactionalId);
+                reenqueue();
+            } else if (error == Errors.COORDINATOR_LOAD_IN_PROGRESS) {
+                reenqueue();
+            } else {
+                result.setError(error.exception());
+            }
+
+            if (error == Errors.NONE || !result.isSuccessful())
+                result.done();
+        }
+
+        @Override
+        public void reenqueue() {
+            pendingTransactionalRequests.add(initPidRequest(true, result));
+        }
+    }
+
+    private class AddPartitionsToTransactionCallback extends TransactionalRequestCallBack {
+
+        AddPartitionsToTransactionCallback() {
+            super(null);
+        }
+
+        @Override
+        public void handleResponse(AbstractResponse response) {
+            AddPartitionsToTxnResponse addPartitionsToTxnResponse = (AddPartitionsToTxnResponse) response;
+            Errors error = addPartitionsToTxnResponse.error();
+            if (error == Errors.NONE) {
+                partitionsInTransaction.addAll(pendingPartitionsToBeAddedToTransaction);
+                pendingPartitionsToBeAddedToTransaction.clear();
+            } else if (error == Errors.COORDINATOR_NOT_AVAILABLE || error == Errors.NOT_COORDINATOR) {
+                needsCoordinator(FindCoordinatorRequest.CoordinatorType.TRANSACTION, transactionalId);
+                reenqueue();
+            } else if (error == Errors.COORDINATOR_LOAD_IN_PROGRESS) {
+                reenqueue();
+            } else if (error == Errors.INVALID_PID_MAPPING || error == Errors.INVALID_TXN_STATE) {
+                throw error.exception();
+            } else if (error == Errors.INVALID_PRODUCER_EPOCH) {
+                isFenced = true;
+                throw error.exception();
```

Comment:
See comment above. We cannot generally raise exceptions from completion handlers.